### ISSUE TITLE
Assign edge feature tensor in HeteroData do not update the metadata.

### ISSUE THREAD

**Issue by igormorgado** [pyg-team/pytorch_geometric#8120]
### 🐛 Describe the bug

For some reason I'm not being able to assign features to edges or they are not correctly updated using [`HeteroData`][1] from PyGeometric package. 

Given the following code snippet

```python
import torch
from torch_geometric.data import HeteroData

data = HeteroData()

# Params
num_papers, num_paper_features = 5, 7
num_authors, num_author_features = 4, 10
num_edges = torch.randint(5, num_papers*num_authors, [1]).item()
author_writes_paper_num_features = 4

# Adding features to nodes
data['paper'].x = torch.randn(num_papers, num_paper_features)
data['author'].x = torch.randn(num_authors, num_author_features)

# Creating some random edges
author_edge_index = torch.randint(0, num_authors, [num_edges])
paper_edge_index = torch.randint(0, num_papers, [num_edges])
edge_index = torch.stack((author_edge_index, paper_edge_index))
data['author', 'writes', 'paper'].edge_index = edge_index

data = data.coalesce()

# Adding features to edges
data['author', 'writes', 'paper'].x = torch.randn(author_writes_paper_num_features, data.num_edges)

# Also tried using the transpose edge feature matrix 
# data['author', 'writes', 'paper'].x = torch.randn(data.num_edges, author_writes_paper_num_features)
```

But, while I can correctly fetch node features

```
data.num_node_features
# {'paper': 7, 'author': 10}

data.node_stores
# [{'x': tensor([[...]])},
#  {'x': tensor([[...]])}]

data.node_attrs()
# ['x']
```

The same does not happens with edge features

```
data.num_edge_features
# {('author', 'writes', 'paper'): 0}
```

Is *ZERO*. Even with a `x` entry in edge_stores.

```
data.edge_stores
#[{'edge_index': tensor([[0, 0, 0, 1, 1, 2, 2, 3, 3],
#         [0, 2, 4, 0, 1, 0, 3, 0, 2]]), 'x': tensor([[...]])}]
```

We can see that no feature attribute `x` is associated with edges

```python
data.edge_attrs()
# ['edge_index']
```

This really seems a bug to me, but it can be only wrong use of the package.

  [1]: https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.data.HeteroData.html#torch_geometric.data.HeteroData

### Environment

* PyG version: 2.3.1
* PyTorch version: 2.0.1+cu117
* OS: Arch Linux
* Python version: 3.11.5
* CUDA/cuDNN version: CUDA 12.2  / cuDNN: 8.9.2.26-1
* How you installed PyTorch and PyG (`conda`, `pip`, source): pip (in venv)
* Any other relevant information (*e.g.*, version of `torch-scatter`): I don't think so

**Comment by igormorgado** [comment]
Talking at PyG slack, I was informed to use `edge_attr` instead `x`.  Using this approach worked as expected. **BUT** there are still some issues here.

1. First and foremost, if this is working as intended, this isn't mentioned at **HeteroData** documentation in API. I could not find a single entry referring to `edge_attr`, but some referring to `edge_attrs()`. 

2. After assign the tensor to the Python attribute `edge_attr`, the HeteroData object correctly returns the the count of features in `.num_edge_features` attribute.

3. Is there any reason to not use the same convention `x` for the feature tensor and `y` to the feature label in edges, as same it's done with the nodes?

4. If I want to assign a label to a edge, similar approach as  node feature `y`s. What is the correct approach? I could also not find anything in documentation.

5. And last, but not least. If for some reason I want to be able to work with graph duals where sometimes nodes and edges interchange, how should I keep this data into the graph? Any convention or best approaches to this using HeteroData (and PyG at all)?


For the reasons above stated I think that is a bug that should be targeted, initially by fixing the documentation, then later, I think, to use the same namings and conventions to edges as we do to nodes. I know that this would break compatibility, but I think is possible to handle both cases during some period, then drop the ``wrong'' way (in my humble opinion) in future major update.


Best regards.

**Comment by igormorgado** [comment]
[Citing Jintang Li:](https://torchgeometricco.slack.com/archives/C01DN0B3B1N/p1696388461373709?thread_ts=1696349473.042079&cid=C01DN0B3B1N)

For edge features, you should use the `edge_attr` rather than `x`.

**Comment by rusty1s** [comment]
Sorry for late response, copy-pasting my answer from Slack:

I cannot reproduce your issue with `edge_attrs()` - this works as intended for me. Can you confirm that `edge_attr.size(0) == edge_index.size(1)`? Otherwise, this attribute will not be classified as an edge-level attribute.

The reason we did go for `edge_attr` is more historically and due to the way `Data` for homogeneous graphs is defined (where we don't have separate storages for node-level and edge-level attributes). Note that there is no problem to define `data[edge_type].x = ...`, it is just that a few methods in PyG (such as `num_edge_features`) expect certain attributes to be set in order to compute it. In the end, you can also do `data[edge_type].x.size(1)` to get the number of edge features.

In general, defining `data.x` and `data.y` is just a convention, it is not a must (and that's why it is not that well documented). You can also store your labels in, e.g., `data.target` or `data.label` and everything should work just fine.

Can you clarify for me how graph duals look like in this case and how you want to use them?
